Support skipping specific rule(s) for a specific task

awcrosby opened the pull request:
Closes #364

Rickkwa commented on README.rst:
```diff
@@ -351,35 +351,43 @@ It's also possible to skip specific rules by passing the rule ID. For example, t
 
-Use the ``warn`` parameter with the *command* or *shell* module.
-
-Use ``skip_ansible_lint`` tag with any task that should be skipped.
```
All documentation for skip_ansible_lint is being removed? To my understanding, the feature is still there, no?

awcrosby replied:
It is not being removed (but may be a good candidate for deprecation in the future). I will update the docs to say it is a less-preferred method to skip.

webknjaz commented on lib/ansiblelint/__init__.py:
```diff
@@ -50,53 +50,76 @@ def matchlines(self, file, text):
+                           file['path'], self, message))
         return matches
 
     def matchtasks(self, file, text):
```
Not strictly related to the current change, but `file` should be `file_` or something that doesn't shadow builtins.

awcrosby replied:
Opened #469 to address this

webknjaz commented on lib/ansiblelint/__init__.py:
```diff
@@ -50,53 +50,76 @@ def matchlines(self, file, text):
+            return matches
+
+        for task in ansiblelint.utils.get_normalized_tasks(yaml, file):
+            if 'skipped_rules' in task and self.id in task['skipped_rules']:
```
```suggestion
            if self.id in task.get('skipped_rules', ()):
```

webknjaz commented on lib/ansiblelint/__init__.py:
```diff
@@ -50,53 +50,76 @@ def matchlines(self, file, text):
+            if not result:
+                continue
+            message = None
+            if isinstance(result, str):
```
This thing is inconsistent with similar check below where `six.string_types` is used. Needs python compat check.

webknjaz commented on lib/ansiblelint/__init__.py:
```diff
@@ -50,53 +50,76 @@ def matchlines(self, file, text):
+            message = None
+            if isinstance(result, six.string_types):
+                message = result
+            taskstr = "Task/Handler: " + ansiblelint.utils.task_to_str(task)
```
snake_case plz
```suggestion
            task_msg = "Task/Handler: " + ansiblelint.utils.task_to_str(task)
```

webknjaz commented on lib/ansiblelint/__init__.py:
```diff
@@ -50,53 +50,76 @@ def matchlines(self, file, text):
+            yaml = [yaml]
+
+        for play in yaml:
+            if 'skipped_rules' in play and self.id in play['skipped_rules']:
```
Same as above: use `.get()`

webknjaz commented on lib/ansiblelint/__init__.py:
```diff
@@ -50,53 +50,76 @@ def matchlines(self, file, text):
+                result = [result]
+
+            if not isinstance(result, list):
+                raise Exception("{} is not a list".format(result))
```
We need to be more precise for exceptions to be useful:
```suggestion
                raise TypeError("{} is not a list".format(result))
```

webknjaz commented on lib/ansiblelint/utils.py:
```diff
@@ -573,3 +575,82 @@ def construct_mapping(node, deep=False):
+        return orig_file_text
+
+    # dump to file with new skipped_rules list(s), read file and return text
+    with tempfile.NamedTemporaryFile('w') as tf:
```
Use in-memory file-like buffers instead.

webknjaz replied:
You can rely on `six.StringIO` (or `six.BytesIO`?)

It should look like:
```python
import six
with six.StringIO() as in_memory_file:
    yaml.dump(data, in_memory_file)
    in_memory_file.seek(0)
    new_file_text = in_memory_file.read()
```

(needs testing)

awcrosby replied:
`io.BytesIO()` worked in py2 and py3

webknjaz commented on lib/ansiblelint/utils.py:
```diff
@@ -573,3 +575,82 @@ def construct_mapping(node, deep=False):
+
+def add_skipped_rules(orig_file_text, file_type):
+    yaml = ruamel.yaml.YAML()
+    data = yaml.load(orig_file_text)
```
load or load`s`?

willthames replied:
Also, should we use safe_load?

awcrosby replied:
`ruamel.yaml` uses `load` instead of `loads`

We are using `yaml = ruamel.yaml.YAML()` and their [docs](https://yaml.readthedocs.io/en/latest/dumpcls.html#dumping-python-classes) state:
> Only yaml = YAML(typ='unsafe') loads and dumps Python objects out-of-the-box. And since it loads any Python object, this can be unsafe.

webknjaz commented on lib/ansiblelint/utils.py:
```diff
@@ -573,3 +575,82 @@ def construct_mapping(node, deep=False):
+    def traverse_yaml(obj):
+        task_comment_obj_strs.append(str(obj.ca.items))
+        if isinstance(obj, dict):
+            for k in obj:
```
```suggestion
            for k, v in obj.items():
```

webknjaz commented on lib/ansiblelint/utils.py:
```diff
@@ -573,3 +575,82 @@ def construct_mapping(node, deep=False):
+                if isinstance(e, (dict, list)):
+                    traverse_yaml(e)
+        else:
+            return
```
This returns `None`. Shouldn't it return `obj`?

awcrosby replied:
The recursive method takes `obj`, but appends to the list defined outside the method, so returning `None` is expected.

ssbarnea commented:
@webknjaz There are not linters supporting noqa for yaml files so far, so we are safe. Even if there would be a collision us very unlikely to cause problems, worst case would be to disable the same rule on two linters. 

I am more than happy to see the same syntax being used. The 2nd linter would have to play nice with ansible-lint not the other way around ;)

Most of the time noqa is used without parameters so no real reasons to worry. I am also worried that any longer requirements would cripple the UX because you may exceed the max line lengh. `  # noqa` adds 8 chars.

willthames commented on setup.cfg:
```diff
@@ -83,6 +83,7 @@ install_requires =
     ansible
     pyyaml
     six
+    ruamel.yaml
```
What is the reliance on ruamel for?

awcrosby replied:
These 2 steps were added between existing ansible-lint functionality
1. (existing) `RulesCollection.run` reads yml file into text
2. (new) `ruamel.yaml` parses text while retaining access to comments through it’s python object
3. (new) `ansiblelint.utils.add_skipped_rules` looks at those comments for any that match a rule skip, if found it adds a `skipped_rules` key to the task with a list of rule ids. `ruamel.yaml` then dumps it’s serialized yaml back into text
4. (existing) Pass text into each rule’s match method such as `matchtasks` which parses using the `yaml` and `ansible` packages (this pr doesn’t try to change any of that parsing)

We are using `yaml = ruamel.yaml.YAML()` and their [docs](https://yaml.readthedocs.io/en/latest/dumpcls.html#dumping-python-classes) state:
> Only yaml = YAML(typ='unsafe') loads and dumps Python objects out-of-the-box. And since it loads any Python object, this can be unsafe.

awcrosby commented on test/TestSkipInsideYaml.py:
```diff
@@ -0,0 +1,114 @@
+import unittest
```
A refactor of ansible-lint test to use pytest overall would probably be a good exercise

awcrosby commented:
Rebased and fixed 2 issues found in this PR:
1. The previous implementation was not retaining line numbers since it was adding to yaml text, now it appends skipped_rules to the pyyaml objects
2. On playbook, fixed to look for skipped rules in all plays
